.github/workflows/on_release.yml

```diff
@@ -7,33 +7,104 @@ on:
   workflow_dispatch:
 
 jobs:
-  docker:
-    strategy:
-      matrix:
-        target: [client, client-standalone, server]
+  docker_release:
     runs-on: ubuntu-latest
     steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
       - name: Set up QEMU
         uses: docker/setup-qemu-action@v1
 
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v1
 
+      - name: Build client
+        id: build_client
+        uses: docker/build-push-action@v2
+        with:
+          file: docker/Dockerfile
+          push: false
+          target: client
+          cache-from: type=registry,ref=authorizon/opal-client:latest
+          cache-to: type=inline
+          tags: |
+            authorizon/opal-client:latest
+            authorizon/opal-client:${{ github.event.release.tag_name }}
+
+      - name: Build client-standalone
+        id: build_client_standalone
+        uses: docker/build-push-action@v2
+        with:
+          file: docker/Dockerfile
+          push: false
+          target: client-standalone
+          cache-from: type=registry,ref=authorizon/opal-client-standalone:latest
+          cache-to: type=inline
+          tags: |
+            authorizon/opal-client-standalone:latest
+            authorizon/opal-client-standalone:${{ github.event.release.tag_name }}
+
+      - name: Build server
+        id: build_server
+        uses: docker/build-push-action@v2
+        with:
+          file: docker/Dockerfile
+          push: false
+          target: server
+          cache-from: type=registry,ref=authorizon/opal-server:latest
+          cache-to: type=inline
+          tags: |
+            authorizon/opal-server:latest
+            authorizon/opal-server:${{ github.event.release.tag_name }}
+
+      - name: Bring up stack
+        run: docker-compose -f docker/docker-compose-example.yml up -d
```

Comment:
i am not sure if this would work. the docker compose takes the images directly from docker hub. i think we should maybe use a slightly different yaml with local tags for it to work.

also - not sure the step actually outputs the built image on the runner machine - maybe the output should be exposed somehow.

i can try to run this locally with `act` without publishing and see what happens.

WDYT @pujan?

Reply:
GitHub Actions steps are running in same VM and if docker has images with same tag it would not pull images from docker hub so I think that shouldn't be an issue, but to be safe we can create different tags which are used in different docker-compose file.

Reply:
i don't know why, but it looks like the build image is indeed overriden by the image from docker hub:

```
Build server...
...
#21 exporting cache
#21 sha256:2700d4ef94dee473593c5c614b55b2dedcca7893909811a8f2b48291a1f581e4
...
Bring up stack...
...
Pulling opal_server (authorizon/opal-server:latest)...
latest: Pulling from authorizon/opal-server
Digest: sha256:f46c7fb76cbefab6b9d5526af318fe0c489a93058890b076c8054e9cf6a4762d
Status: Downloaded newer image for authorizon/opal-server:latest
Pulling opal_client (authorizon/opal-client:latest)...
latest: Pulling from authorizon/opal-client
Digest: sha256:3c5d6fd589a29285aa852165a2183f62a086773be7a84b8fa34c4c042d89ed10
Status: Downloaded newer image for authorizon/opal-client:latest
```
different hash is used - i also inserted a dummy log line to make sure which image is used.

no worries, i'm fixing it.

Reply:
run link: https://github.com/authorizon/opal/runs/2631173231?check_suite_focus=true

Reply:
ok i figured it out: github action "docker/build-push-action@v2" does not store the images on the runner machine, it caches them somewhere else, and that is why the runner machine fetches the containers from docker hub instead. i couldn't figure out easily how to fix that, so i created a workaround.